lib/util/graceful_shutdown.c

```diff
@@ -0,0 +1,101 @@
+#include <signal.h>
+
+#include "events.h"
+#include "warnp.h"
+
+#include "graceful_shutdown.h"
+
+/* Data from parent code. */
+static int (* begin_shutdown)(void *);
+static void (* sighandler_sigterm_orig)(int);
+static void * caller_cookie;
+static void * timer_cookie;
+
+/* Flag to show that SIGTERM was given. */
```

Comment:
What we *care about* is whether it was sent.  But what we *know about* is whether it was received. :-)